lib/Parse/ParseExpr.cpp

```diff
@@ -2519,9 +2519,6 @@ ParserStatus Parser::parseClosureSignatureIfPresent(
       }
     }
     
-    // Parse the 'in' at the end.
-    if (Tok.isNot(tok::kw_in))
-      return makeParserSuccess();
```

Comment:
@kaar3k Thanks for looking into this!
However, this is too aggressive. See the following example:

```swift
_ = { @Wrapper let abc = 1 }
_ = { [foo, bar, baz].map { process($0) } }
_ = { call(value) }
_ = { _ = someThing() }
_ = { (foo, bar) = returnTuple() }
```

They are all valid closures and we need to keep them valid. But your change break them because the first part is ambiguous without `in` keyword.
So we need some heuristics. For example

* `[...]` followed by `.`, `[`, etc. is probably an expression, not capture list.
* `(...)` followed by `.`, `=`, etc. is probably an expression, not parameter list.

And if we are unsure, we should parse them as the body not as the signature, so that we don't break any existing code.

Reply:
Thank you very much for the detailed context @rintaro , this makes sense..  Totally agree with your point parsing them as part of body. let me come up with right fix for this,